{
 "cells": [
  {
   "cell_type": "code",
   "id": "e080f1f0d3be29e4",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "metadata = {\n",
    "    'Author      ': 'Jay Annadurai',\n",
    "    'Date        ': '12 Apr 2024',\n",
    "    'Project     ': 'A4-PSWMPredictor',\n",
    "    'Version     ': 1.0,\n",
    "    'Description ': 'ArgR Binding Site Motif Predictor with PSWM'\n",
    "}"
   ],
   "outputs": [],
   "execution_count": 178
  },
  {
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~\n",
    "#  Import Libraries\n",
    "# ~~~~~~~~~~~~~~~~~~\n",
    "import pandas as pd  # Data Reading\n",
    "import numpy as np  # Computation\n",
    "import scipy as sp # Statistical Methods\n",
    "import seaborn as sb  # Advanced Data Visualization\n",
    "import matplotlib.pyplot as plt  # Data Visualization\n",
    "# Set Matplot to Inline\n",
    "%matplotlib inline \n",
    "\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Import Utlity Classes\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~\n",
    "from pprint import pprint as print  # Override the standard print function with Pretty Print\n",
    "from JayUtilities import DataIO as Jio  # Data Input/Output Processing Utility Class"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "1ef5aa2289942f9d",
   "outputs": [],
   "execution_count": 179
  },
  {
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~\n",
    "#  Script Config\n",
    "# ~~~~~~~~~~~~~~~\n",
    "# Input Config\n",
    "Jio.input_folder = \"Input/\"  # Sets the Input Folder for the DataIO Class\n",
    "input_file_paths = { # Name of the Input File\n",
    "    'ArgR-Motif': 'argR-counts-matrix.txt',\n",
    "    'Ecoli-Genetic-Regions':'E_coli_K12_MG1655.400_50'\n",
    "}\n",
    "\n",
    "input_format = 'tsv'  # Optionally force the encode format of the Input File\n",
    "header_rows = 0  # Ptt files contain 2 rows of header information\n",
    "\n",
    "\n",
    "# Output Config\n",
    "Jio.output_folder = \"Output/\"  # Sets the Output Folder for the DataIO Class\n",
    "save_file = True  # Sets whether the script should save the outputs or not\n",
    "output_file = None  # Name of the Output File is generated Dynamically\n",
    "output_format = 'tsv'  # Format of the file to save the Output as"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "3cc80185d8e8517a",
   "outputs": [],
   "execution_count": 180
  },
  {
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Import Counts Matrix Data\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "# Import the File\n",
    "counts_matrix_dict = Jio.file_to_df(\n",
    "    file_name=input_file_paths['ArgR-Motif'], \n",
    "    force_encode_format=input_format,\n",
    "    read_args= { \"header\": None} \n",
    ")\n",
    "\n",
    "# Bind the Dataframe\n",
    "raw_counts_matrix_df = counts_matrix_dict['df']\n",
    "\n",
    "# View Data\n",
    "Jio.print_df(raw_counts_matrix_df, df_name=\"ArgR Binding Motif Counts Matrix Raw\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4230f3c268f73572",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "ArgR Binding Motif Counts Matrix Raw: 4 Row x 20 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "  0  1   2   3   4   5   6   7   8   9   10  11  12  13  14  15  16  17  18  \\\n",
      "0  a  |   8  12  21   9   4   2  21  21   3  10   8   5   7  25   4   2   2   \n",
      "1  c  |   7   4   1   6   2   3   3   3   1   0   2   0   7   0   3   3  24   \n",
      "2  g  |   3   2   1   8   2  21   2   2   0   1   0   1   0   1   0  15   0   \n",
      "3  t  |   9   9   4   4  19   1   1   1  23  16  17  21  13   1  20   7   1   \n",
      "\n",
      "   19  \n",
      "0  25  \n",
      "1   0  \n",
      "2   2  \n",
      "3   0  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 181
  },
  {
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Counts Matrix Preprocessing\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Copy DF to prevent any code issues\n",
    "counts_matrix_df = raw_counts_matrix_df.copy()\n",
    "\n",
    "# Drop the original index and use the base as the new index\n",
    "counts_matrix_df.set_index(0, inplace=True)\n",
    "# Drop the formatting column\n",
    "counts_matrix_df.drop(1,axis=1,inplace=True)\n",
    "\n",
    "# Renaming columns to P1 to P18 to represent the position\n",
    "new_column_names = []\n",
    "for i in range(1, len(counts_matrix_df.columns) + 1):\n",
    "    new_column_name = f\"P{i}\"\n",
    "    new_column_names.append(new_column_name)\n",
    "\n",
    "counts_matrix_df.columns = new_column_names\n",
    "\n",
    "# View Data\n",
    "Jio.print_df(counts_matrix_df, df_name=\"ArgR Binding Motif Counts Matrix\")"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2a442e74fb9e5445",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "ArgR Binding Motif Counts Matrix: 4 Row x 18 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "   P1  P2  P3  P4  P5  P6  P7  P8  P9  P10  P11  P12  P13  P14  P15  P16  P17  \\\n",
      "0                                                                               \n",
      "a   8  12  21   9   4   2  21  21   3   10    8    5    7   25    4    2    2   \n",
      "c   7   4   1   6   2   3   3   3   1    0    2    0    7    0    3    3   24   \n",
      "g   3   2   1   8   2  21   2   2   0    1    0    1    0    1    0   15    0   \n",
      "t   9   9   4   4  19   1   1   1  23   16   17   21   13    1   20    7    1   \n",
      "\n",
      "   P18  \n",
      "0       \n",
      "a   25  \n",
      "c    0  \n",
      "g    2  \n",
      "t    0  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 182
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Counts Matrix with Psuedocount\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Function takes a counts matrix df and returns it with the psuedo count applied\n",
    "def add_psuedocount(counts_matrix_df, psuedocount=1):\n",
    "    # Add one to all positions in the counts matrix\n",
    "    pseudocounts_matrix_df = counts_matrix_df + psuedocount\n",
    "    return pseudocounts_matrix_df"
   ],
   "id": "36f6526c0f3b60ce",
   "outputs": [],
   "execution_count": 183
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Nucleotide Frequency Matrix\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Computes the Frequency Matrix from the Counts Matrix\n",
    "def counts_to_frequency_matrix_df(counts_matrix_df):\n",
    "    # Calculate the sum of each column (each position in the motif)\n",
    "    column_sums = counts_matrix_df.sum()\n",
    "\n",
    "    # Divide each element in the DataFrame by its column sum to get frequencies\n",
    "    frequency_matrix_df = counts_matrix_df.div(column_sums, axis='columns')\n",
    "\n",
    "    # Return the Frequency Matrix\n",
    "    return frequency_matrix_df\n",
    "    "
   ],
   "id": "e89a28508ea93fe",
   "outputs": [],
   "execution_count": 184
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Apply and View the Psuedocount Transformation\n",
    "psuedocounts_matrix_df = add_psuedocount(counts_matrix_df)\n",
    "\n",
    "Jio.print_df(psuedocounts_matrix_df,df_name=\"ArgR Binding Motif Psuedocounts Matrix\")"
   ],
   "id": "684a4d1ce834c424",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "ArgR Binding Motif Psuedocounts Matrix: 4 Row x 18 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "   P1  P2  P3  P4  P5  P6  P7  P8  P9  P10  P11  P12  P13  P14  P15  P16  P17  \\\n",
      "0                                                                               \n",
      "a   9  13  22  10   5   3  22  22   4   11    9    6    8   26    5    3    3   \n",
      "c   8   5   2   7   3   4   4   4   2    1    3    1    8    1    4    4   25   \n",
      "g   4   3   2   9   3  22   3   3   1    2    1    2    1    2    1   16    1   \n",
      "t  10  10   5   5  20   2   2   2  24   17   18   22   14    2   21    8    2   \n",
      "\n",
      "   P18  \n",
      "0       \n",
      "a   26  \n",
      "c    1  \n",
      "g    3  \n",
      "t    1  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 185
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Apply and View the Frequency Transformation\n",
    "frequency_matrix_df = counts_to_frequency_matrix_df(psuedocounts_matrix_df)\n",
    "\n",
    "Jio.print_df(frequency_matrix_df,df_name=\"ArgR Binding Motif Frequency Matrix (with Psuedocount)\")"
   ],
   "id": "e434b77210b37a86",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "ArgR Binding Motif Frequency Matrix (with Psuedocount): 4 Row x 18 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "         P1        P2        P3        P4        P5        P6        P7  \\\n",
      "0                                                                         \n",
      "a  0.290323  0.419355  0.709677  0.322581  0.161290  0.096774  0.709677   \n",
      "c  0.258065  0.161290  0.064516  0.225806  0.096774  0.129032  0.129032   \n",
      "g  0.129032  0.096774  0.064516  0.290323  0.096774  0.709677  0.096774   \n",
      "t  0.322581  0.322581  0.161290  0.161290  0.645161  0.064516  0.064516   \n",
      "\n",
      "         P8        P9       P10       P11       P12       P13       P14  \\\n",
      "0                                                                         \n",
      "a  0.709677  0.129032  0.354839  0.290323  0.193548  0.258065  0.838710   \n",
      "c  0.129032  0.064516  0.032258  0.096774  0.032258  0.258065  0.032258   \n",
      "g  0.096774  0.032258  0.064516  0.032258  0.064516  0.032258  0.064516   \n",
      "t  0.064516  0.774194  0.548387  0.580645  0.709677  0.451613  0.064516   \n",
      "\n",
      "        P15       P16       P17       P18  \n",
      "0                                          \n",
      "a  0.161290  0.096774  0.096774  0.838710  \n",
      "c  0.129032  0.129032  0.806452  0.032258  \n",
      "g  0.032258  0.516129  0.032258  0.096774  \n",
      "t  0.677419  0.258065  0.064516  0.032258  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 186
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Log Odds Ratio Frequency Matrix\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Computes matrix of the log odds ratio between frequency and expected frequency\n",
    "# Expected P is the probability a nucleotide should appear (base is 1 / 4)\n",
    "# Background is set to 1\n",
    "def frequency_to_log_odds_matrix_df(frequency_matrix_df, expected_probability = (1/4), background = 1, frequency_uses_pseudocount=True):\n",
    "\n",
    "    # If the frequency matrix already adjusts via psuedocount\n",
    "    if frequency_uses_pseudocount:\n",
    "        # Formula is Ln ( Frequency / Expected Frequency ) / Background\n",
    "        log_odds_matrix_df = (\n",
    "                np.log( frequency_matrix_df / expected_probability )\n",
    "                /\n",
    "                background\n",
    "        )\n",
    "        \n",
    "    # If the psuedocount is not previously adjusted, add a minor amount to prevent ln(0)\n",
    "    else:\n",
    "        # Set an insignificant adjustment to prevent division log of zero\n",
    "        adjustment = 1e-100\n",
    "\n",
    "        # Return the Log Odds Matrix with an Adjustment Factor\n",
    "        log_odds_matrix_df = (\n",
    "            np.log(\n",
    "                (frequency_matrix_df + adjustment) \n",
    "                / \n",
    "                (expected_probability + adjustment)\n",
    "            ) \n",
    "            /\n",
    "            background\n",
    "        )\n",
    "\n",
    "    # Return the Log Odds Matrix DF\n",
    "    return log_odds_matrix_df"
   ],
   "id": "a9157d599ec271ea",
   "outputs": [],
   "execution_count": 187
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Apply and View the Log Odds Ratio Transformation\n",
    "log_odds_matrix_df = frequency_to_log_odds_matrix_df(frequency_matrix_df)\n",
    "\n",
    "Jio.print_df(log_odds_matrix_df,df_name=\"ArgR Binding Motif Log Odds Frequency Matrix\")"
   ],
   "id": "9ced15f9790220ca",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "ArgR Binding Motif Log Odds Frequency Matrix: 4 Row x 18 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "         P1        P2        P3        P4        P5        P6        P7  \\\n",
      "0                                                                         \n",
      "a  0.149532  0.517257  1.043350  0.254892 -0.438255 -0.949081  1.043350   \n",
      "c  0.031749 -0.438255 -1.354546 -0.101783 -0.949081 -0.661398 -0.661398   \n",
      "g -0.661398 -0.949081 -1.354546  0.149532 -0.949081  1.043350 -0.949081   \n",
      "t  0.254892  0.254892 -0.438255 -0.438255  0.948039 -1.354546 -1.354546   \n",
      "\n",
      "         P8        P9       P10       P11       P12       P13       P14  \\\n",
      "0                                                                         \n",
      "a  1.043350 -0.661398  0.350202  0.149532 -0.255933  0.031749  1.210404   \n",
      "c -0.661398 -1.354546 -2.047693 -0.949081 -2.047693  0.031749 -2.047693   \n",
      "g -0.949081 -2.047693 -1.354546 -2.047693 -1.354546 -2.047693 -1.354546   \n",
      "t -1.354546  1.130361  0.785521  0.842679  1.043350  0.591364 -1.354546   \n",
      "\n",
      "        P15       P16       P17       P18  \n",
      "0                                          \n",
      "a -0.438255 -0.949081 -0.949081  1.210404  \n",
      "c -0.661398 -0.661398  1.171183 -2.047693  \n",
      "g -2.047693  0.724896 -2.047693 -0.949081  \n",
      "t  0.996830  0.031749 -1.354546 -2.047693  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 188
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Position-Specific Weight|Scoring Matrix (PSWM or PSSM)\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Function combines the stages of generating a PSW|PSS Matrix\n",
    "def count_matrix_to_pswm_df(counts_matrix_df):\n",
    "    \n",
    "    # Get the Psuedocount Matrix\n",
    "    psuedocounts_matrix_df = add_psuedocount(counts_matrix_df)\n",
    "    \n",
    "    # Get the Frequency Matrix from the Counts Matrix with Psuedocount\n",
    "    psuedocount_frequency_matrix_df = counts_to_frequency_matrix_df(psuedocounts_matrix_df)\n",
    "    \n",
    "    # Get the Log Odds Ratio Frequency Matrix from the Frequency Matrix\n",
    "    pswm_df = frequency_to_log_odds_matrix_df(psuedocount_frequency_matrix_df, expected_probability = (1/4), background = 1, frequency_uses_pseudocount = True)\n",
    "    \n",
    "    # Return the Position Specific Weight|Scoring Matrix\n",
    "    return pswm_df"
   ],
   "id": "8b973ac0e1b1a9a2",
   "outputs": [],
   "execution_count": 189
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# One-step conversion from the Counts Matrix to the PSWM Matrix\n",
    "pswm_df = count_matrix_to_pswm_df(counts_matrix_df)\n",
    "\n",
    "Jio.print_df(log_odds_matrix_df,df_name=\"ArgR Binding Motif PSWM\")"
   ],
   "id": "4dff9fb1dddb1ace",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "ArgR Binding Motif PSWM: 4 Row x 18 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "         P1        P2        P3        P4        P5        P6        P7  \\\n",
      "0                                                                         \n",
      "a  0.149532  0.517257  1.043350  0.254892 -0.438255 -0.949081  1.043350   \n",
      "c  0.031749 -0.438255 -1.354546 -0.101783 -0.949081 -0.661398 -0.661398   \n",
      "g -0.661398 -0.949081 -1.354546  0.149532 -0.949081  1.043350 -0.949081   \n",
      "t  0.254892  0.254892 -0.438255 -0.438255  0.948039 -1.354546 -1.354546   \n",
      "\n",
      "         P8        P9       P10       P11       P12       P13       P14  \\\n",
      "0                                                                         \n",
      "a  1.043350 -0.661398  0.350202  0.149532 -0.255933  0.031749  1.210404   \n",
      "c -0.661398 -1.354546 -2.047693 -0.949081 -2.047693  0.031749 -2.047693   \n",
      "g -0.949081 -2.047693 -1.354546 -2.047693 -1.354546 -2.047693 -1.354546   \n",
      "t -1.354546  1.130361  0.785521  0.842679  1.043350  0.591364 -1.354546   \n",
      "\n",
      "        P15       P16       P17       P18  \n",
      "0                                          \n",
      "a -0.438255 -0.949081 -0.949081  1.210404  \n",
      "c -0.661398 -0.661398  1.171183 -2.047693  \n",
      "g -2.047693  0.724896 -2.047693 -0.949081  \n",
      "t  0.996830  0.031749 -1.354546 -2.047693  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 190
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Save the PSWM for the Arg-R Binding Motif Region\n",
    "if save_file:\n",
    "    \n",
    "    # The PSWM DF to Save includes the Index as a Column\n",
    "    pswm_df_to_save = pswm_df.copy().reset_index(drop=False)\n",
    "    \n",
    "    # Save the DF in the specified file format\n",
    "    Jio.df_to_file(\n",
    "        df=pswm_df_to_save,\n",
    "        file_name=f\"ArgR-Motif-PSWM\",\n",
    "        file_format=output_format\n",
    "    )"
   ],
   "id": "8344ff4be3f26b42",
   "outputs": [],
   "execution_count": 191
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Import Ecoli Genetic Regions Data\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "# Import the File\n",
    "genetic_regions_dict = Jio.file_to_df(\n",
    "    file_name=input_file_paths['Ecoli-Genetic-Regions'],\n",
    "    force_encode_format=input_format,\n",
    "    read_args= { \"header\": None}\n",
    ")\n",
    "\n",
    "# Bind the Dataframe\n",
    "raw_genetic_regions_df = genetic_regions_dict['df']\n",
    "\n",
    "# View Data\n",
    "Jio.print_df(raw_genetic_regions_df, df_name=\"E coli Genetic Regions Raw\")"
   ],
   "id": "733673b5db088a20",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "E coli Genetic Regions Raw: 4319 Row x 1 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "                                                   0\n",
      "0  16127995 \\ aacggcagaccaacatcaactgcaagctttacgcg...\n",
      "1  16127996 \\ ccgtttgctgcatgatattgaaaaaaatatcacca...\n",
      "2  16127997 \\ gaaacgggacgtgaactggagctggcggatattga...\n",
      "3  16127998 \\ ggggattaaagtctcgacggcagaagccagggcta...\n",
      "4  16127999 \\ aggcgaatatggcttgttcctcggcaccgcgcatc...\n",
      "5  16128000 \\ ctacatttatgccgaaaacaatctcttctttttac...\n",
      "6  16128001 \\ ttaagaatgagagaagggttggttgtggcatcctg...\n",
      "7  16128002 \\ cggatgtagcgaaactgcacaaatccggtgcgaaa...\n",
      "8  16128003 \\ gggcgcaagcttccgtaacatcggcgaaattctgg...\n",
      "9  16128004 \\ ctgatgaaagacaaattattggcaaaaggtgtgtc...\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 192
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Ecoli Genetic Regions Preprocessing\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Copy DF to prevent any code issues\n",
    "genetic_regions_df = raw_genetic_regions_df.copy()\n",
    "\n",
    "# Split the FASTA Nucleotide Sequence and the Gene ID from the single Row value\n",
    "genetic_regions_df[['GeneID', 'Sequence']] = (\n",
    "    genetic_regions_df[0].apply(lambda x: pd.Series(x.split(' \\ ')))\n",
    ")\n",
    "\n",
    "# # Set the GeneID as the Index \n",
    "# genetic_regions_df.set_index('GeneID', inplace=True)\n",
    "\n",
    "# Remove the trailing backslash ' \\' from the 'Sequence' column\n",
    "genetic_regions_df['Sequence'] = genetic_regions_df['Sequence'].str.rstrip(' \\\\')\n",
    "\n",
    "\n",
    "# Split the Upstream and Downstream Regions of the Genetic Region\n",
    "genetic_regions_df['Upstream'] = genetic_regions_df['Sequence'].str.slice(0, 400)\n",
    "genetic_regions_df['Downstream'] = genetic_regions_df['Sequence'].str.slice(400)\n",
    "\n",
    "# Drop the original column if no longer needed\n",
    "genetic_regions_df.drop(0, axis='columns', inplace=True)\n",
    "\n",
    "# View Data\n",
    "Jio.print_df(genetic_regions_df, df_name=\"E coli Genetic Regions\")"
   ],
   "id": "8b2d9fb24f53f80d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "E coli Genetic Regions: 4319 Row x 4 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "     GeneID                                           Sequence  \\\n",
      "0  16127995  aacggcagaccaacatcaactgcaagctttacgcgaacgagccatg...   \n",
      "1  16127996  ccgtttgctgcatgatattgaaaaaaatatcaccaaataaaaaacg...   \n",
      "2  16127997  gaaacgggacgtgaactggagctggcggatattgaaattgaacctg...   \n",
      "3  16127998  ggggattaaagtctcgacggcagaagccagggctattttaccggcg...   \n",
      "4  16127999  aggcgaatatggcttgttcctcggcaccgcgcatccggcgaaattt...   \n",
      "5  16128000  ctacatttatgccgaaaacaatctcttctttttacgcctgaacaac...   \n",
      "6  16128001  ttaagaatgagagaagggttggttgtggcatcctgcggttgataca...   \n",
      "7  16128002  cggatgtagcgaaactgcacaaatccggtgcgaaaagtgaaccaac...   \n",
      "8  16128003  gggcgcaagcttccgtaacatcggcgaaattctggaactggcaggc...   \n",
      "9  16128004  ctgatgaaagacaaattattggcaaaaggtgtgtcgcatttgcttt...   \n",
      "\n",
      "                                            Upstream  \\\n",
      "0  aacggcagaccaacatcaactgcaagctttacgcgaacgagccatg...   \n",
      "1  ccgtttgctgcatgatattgaaaaaaatatcaccaaataaaaaacg...   \n",
      "2  gaaacgggacgtgaactggagctggcggatattgaaattgaacctg...   \n",
      "3  ggggattaaagtctcgacggcagaagccagggctattttaccggcg...   \n",
      "4  aggcgaatatggcttgttcctcggcaccgcgcatccggcgaaattt...   \n",
      "5  ctacatttatgccgaaaacaatctcttctttttacgcctgaacaac...   \n",
      "6  ttaagaatgagagaagggttggttgtggcatcctgcggttgataca...   \n",
      "7  cggatgtagcgaaactgcacaaatccggtgcgaaaagtgaaccaac...   \n",
      "8  gggcgcaagcttccgtaacatcggcgaaattctggaactggcaggc...   \n",
      "9  ctgatgaaagacaaattattggcaaaaggtgtgtcgcatttgcttt...   \n",
      "\n",
      "                                          Downstream  \n",
      "0  atgaaacgcattagcaccaccattaccaccaccatcaccattacca...  \n",
      "1  atgcgagtgttgaagttcggcggtacatcagtggcaaatgcagaac...  \n",
      "2  atggttaaagtttatgccccggcttccagtgccaatatgagcgtcg...  \n",
      "3  atgaaactctacaatctgaaagatcacaacgagcaggtcagctttg...  \n",
      "4  gtgaaaaagatgcaatctatcgtactcgcactttccctggttctgg...  \n",
      "5  atgctgattcttatttcacctgcgaaaacgcttgattaccaaagcc...  \n",
      "6  atgccagattttttctccttcattaacagcgtcctttggggatcgg...  \n",
      "7  atgacggacaaattgacctcccttcgtcagtacaccaccgtagtgg...  \n",
      "8  atgaatactttacgtattggcttagtttccatctctgatcgcgcat...  \n",
      "9  atgggcaacactaagttggctaatccggcaccgctgggcctgatgg...  \n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 193
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Check Lengths of Sequences to ensure they meet the specifications\n",
    "\n",
    "# Copy DF to prevent any code issues\n",
    "region_lengths_df = genetic_regions_df.copy()\n",
    "\n",
    "# Calculate lengths for Sequence, Upstream, and Downstream columns\n",
    "# Should be 450\n",
    "region_lengths_df['Seq_length'] = region_lengths_df['Sequence'].apply(len)\n",
    "# Should be 400\n",
    "region_lengths_df['Upstream_length'] = region_lengths_df['Upstream'].apply(len)\n",
    "# Should be  50\n",
    "region_lengths_df['Downstream_length'] = region_lengths_df['Downstream'].apply(len)\n",
    "\n",
    "# Drop the original columns\n",
    "region_lengths_df.drop(\n",
    "    ['Sequence','Upstream','Downstream'],\n",
    "    axis='columns', inplace=True\n",
    ")\n",
    "\n",
    "# Inpsect the Metadata regarding Length per \n",
    "Jio.print_df(region_lengths_df, df_name=\"E coli Genetic Regions Metadata\",rows=3)"
   ],
   "id": "7a88a89bfd8aa134",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "E coli Genetic Regions Metadata: 4319 Row x 4 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "     GeneID  Seq_length  Upstream_length  Downstream_length\n",
      "0  16127995         450              400                 50\n",
      "1  16127996         450              400                 50\n",
      "2  16127997         450              400                 50\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 194
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "genetic_regions_df",
   "id": "c95d569cd886b8de",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "        GeneID                                           Sequence  \\\n",
       "0     16127995  aacggcagaccaacatcaactgcaagctttacgcgaacgagccatg...   \n",
       "1     16127996  ccgtttgctgcatgatattgaaaaaaatatcaccaaataaaaaacg...   \n",
       "2     16127997  gaaacgggacgtgaactggagctggcggatattgaaattgaacctg...   \n",
       "3     16127998  ggggattaaagtctcgacggcagaagccagggctattttaccggcg...   \n",
       "4     16127999  aggcgaatatggcttgttcctcggcaccgcgcatccggcgaaattt...   \n",
       "...        ...                                                ...   \n",
       "4314  16132216  gtgctgacgactacgtggctaaaccgttttcaccccgcgaagtgtg...   \n",
       "4315  16132217  tgctgctgaaccggcgttactggagcaggcgctgggaaatttactg...   \n",
       "4316  16132218  acgtttattagttgtatgatgcaactagttggattattaaaataat...   \n",
       "4317  16132219  gtgatgtagtcatctgcaccgatttcgaggccgagaattttatcga...   \n",
       "4318  16132220  acaggctaagaggggccggacacccaataaaactacgcttcgttga...   \n",
       "\n",
       "                                               Upstream  \\\n",
       "0     aacggcagaccaacatcaactgcaagctttacgcgaacgagccatg...   \n",
       "1     ccgtttgctgcatgatattgaaaaaaatatcaccaaataaaaaacg...   \n",
       "2     gaaacgggacgtgaactggagctggcggatattgaaattgaacctg...   \n",
       "3     ggggattaaagtctcgacggcagaagccagggctattttaccggcg...   \n",
       "4     aggcgaatatggcttgttcctcggcaccgcgcatccggcgaaattt...   \n",
       "...                                                 ...   \n",
       "4314  gtgctgacgactacgtggctaaaccgttttcaccccgcgaagtgtg...   \n",
       "4315  tgctgctgaaccggcgttactggagcaggcgctgggaaatttactg...   \n",
       "4316  acgtttattagttgtatgatgcaactagttggattattaaaataat...   \n",
       "4317  gtgatgtagtcatctgcaccgatttcgaggccgagaattttatcga...   \n",
       "4318  acaggctaagaggggccggacacccaataaaactacgcttcgttga...   \n",
       "\n",
       "                                             Downstream  \n",
       "0     atgaaacgcattagcaccaccattaccaccaccatcaccattacca...  \n",
       "1     atgcgagtgttgaagttcggcggtacatcagtggcaaatgcagaac...  \n",
       "2     atggttaaagtttatgccccggcttccagtgccaatatgagcgtcg...  \n",
       "3     atgaaactctacaatctgaaagatcacaacgagcaggtcagctttg...  \n",
       "4     gtgaaaaagatgcaatctatcgtactcgcactttccctggttctgg...  \n",
       "...                                                 ...  \n",
       "4314  atgcgtatcggcatgcggttgttgctgggctattttttactggtgg...  \n",
       "4315  atgttgaaatcccccctgttctggaaaatgactagcctgtttggtg...  \n",
       "4316  atgcagaccccgcacattcttatcgttgaagacgagttggtaacac...  \n",
       "4317  atgactaaagtacgtaattgcgttcttgatgcactttccatcaacg...  \n",
       "4318  atgcgtataacgattattctggtcgcacccgccagagcagaaaata...  \n",
       "\n",
       "[4319 rows x 4 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Upstream</th>\n",
       "      <th>Downstream</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16127995</td>\n",
       "      <td>aacggcagaccaacatcaactgcaagctttacgcgaacgagccatg...</td>\n",
       "      <td>aacggcagaccaacatcaactgcaagctttacgcgaacgagccatg...</td>\n",
       "      <td>atgaaacgcattagcaccaccattaccaccaccatcaccattacca...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16127996</td>\n",
       "      <td>ccgtttgctgcatgatattgaaaaaaatatcaccaaataaaaaacg...</td>\n",
       "      <td>ccgtttgctgcatgatattgaaaaaaatatcaccaaataaaaaacg...</td>\n",
       "      <td>atgcgagtgttgaagttcggcggtacatcagtggcaaatgcagaac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>16127997</td>\n",
       "      <td>gaaacgggacgtgaactggagctggcggatattgaaattgaacctg...</td>\n",
       "      <td>gaaacgggacgtgaactggagctggcggatattgaaattgaacctg...</td>\n",
       "      <td>atggttaaagtttatgccccggcttccagtgccaatatgagcgtcg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>16127998</td>\n",
       "      <td>ggggattaaagtctcgacggcagaagccagggctattttaccggcg...</td>\n",
       "      <td>ggggattaaagtctcgacggcagaagccagggctattttaccggcg...</td>\n",
       "      <td>atgaaactctacaatctgaaagatcacaacgagcaggtcagctttg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16127999</td>\n",
       "      <td>aggcgaatatggcttgttcctcggcaccgcgcatccggcgaaattt...</td>\n",
       "      <td>aggcgaatatggcttgttcctcggcaccgcgcatccggcgaaattt...</td>\n",
       "      <td>gtgaaaaagatgcaatctatcgtactcgcactttccctggttctgg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4314</th>\n",
       "      <td>16132216</td>\n",
       "      <td>gtgctgacgactacgtggctaaaccgttttcaccccgcgaagtgtg...</td>\n",
       "      <td>gtgctgacgactacgtggctaaaccgttttcaccccgcgaagtgtg...</td>\n",
       "      <td>atgcgtatcggcatgcggttgttgctgggctattttttactggtgg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4315</th>\n",
       "      <td>16132217</td>\n",
       "      <td>tgctgctgaaccggcgttactggagcaggcgctgggaaatttactg...</td>\n",
       "      <td>tgctgctgaaccggcgttactggagcaggcgctgggaaatttactg...</td>\n",
       "      <td>atgttgaaatcccccctgttctggaaaatgactagcctgtttggtg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4316</th>\n",
       "      <td>16132218</td>\n",
       "      <td>acgtttattagttgtatgatgcaactagttggattattaaaataat...</td>\n",
       "      <td>acgtttattagttgtatgatgcaactagttggattattaaaataat...</td>\n",
       "      <td>atgcagaccccgcacattcttatcgttgaagacgagttggtaacac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4317</th>\n",
       "      <td>16132219</td>\n",
       "      <td>gtgatgtagtcatctgcaccgatttcgaggccgagaattttatcga...</td>\n",
       "      <td>gtgatgtagtcatctgcaccgatttcgaggccgagaattttatcga...</td>\n",
       "      <td>atgactaaagtacgtaattgcgttcttgatgcactttccatcaacg...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4318</th>\n",
       "      <td>16132220</td>\n",
       "      <td>acaggctaagaggggccggacacccaataaaactacgcttcgttga...</td>\n",
       "      <td>acaggctaagaggggccggacacccaataaaactacgcttcgttga...</td>\n",
       "      <td>atgcgtataacgattattctggtcgcacccgccagagcagaaaata...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4319 rows × 4 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 195,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 195
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Prepare the Genetic Regions DF to be evaluated by PSWM\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
   ],
   "id": "1b4a4e44d247a87",
   "outputs": [],
   "execution_count": 196
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Function to Partition a Dataframe by Chunks\n",
    "def partition_df(df, partition_size, partition_col='Sequence', identifier_col=None, partial_partitions=True, overlaps=False):\n",
    "    # Prepare a list to collect the partitioned sequences\n",
    "    partitions = []\n",
    "\n",
    "    # Extract index name if identifier_col is None and index has a name\n",
    "    index_name = df.index.name if identifier_col is None and df.index.name is not None else 'Index'\n",
    "\n",
    "    # Determine the step size based on whether overlaps are allowed\n",
    "    step_size = 1 if overlaps else partition_size\n",
    "\n",
    "    # Iterate through each row in the DataFrame\n",
    "    for index, row in df.iterrows():\n",
    "        # Determine the identifier based on whether a specific identifier column is provided\n",
    "        identifier = row[identifier_col] if identifier_col else index\n",
    "\n",
    "        # Extract the sequence\n",
    "        sequence = row[partition_col]\n",
    "\n",
    "        # Generate partitions of the sequence\n",
    "        for i in range(0, len(sequence) - partition_size + 1, step_size):\n",
    "            partitioned_sequence = sequence[i:i + partition_size]\n",
    "            start_point = i\n",
    "            end_point = i + len(partitioned_sequence) - 1\n",
    "\n",
    "            # Prepare the data tuple\n",
    "            partition_data = (\n",
    "                identifier,\n",
    "                partitioned_sequence,\n",
    "                start_point,\n",
    "                end_point\n",
    "            )\n",
    "\n",
    "            # Add the partition depending on whether partial partitions are allowed\n",
    "            if partial_partitions or len(partitioned_sequence) == partition_size:\n",
    "                partitions.append(partition_data)\n",
    "\n",
    "    # Define column names based on whether an identifier column is provided\n",
    "    columns = [identifier_col if identifier_col else index_name, f'{partition_col}_Partition', 'Start', 'End']\n",
    "\n",
    "    # Create a new DataFrame from the partitions list\n",
    "    long_df = pd.DataFrame(partitions, columns=columns)\n",
    "\n",
    "    # Set the index of the DataFrame if no identifier column was specified\n",
    "    if identifier_col is None:\n",
    "        long_df.set_index(index_name, inplace=True)\n",
    "\n",
    "    # Return the long DataFrame\n",
    "    return long_df"
   ],
   "id": "81d21f93f3d07f45",
   "outputs": [],
   "execution_count": 197
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Partition Each Sequence of the Gene \n",
    "partitioned_genetic_regions_df = partition_df(genetic_regions_df, partition_size=len(pswm_df.columns), identifier_col = 'GeneID', overlaps=True)"
   ],
   "id": "19b135764f296fa8",
   "outputs": [],
   "execution_count": 198
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Generate all the Partitions that Need to be Evaluated\n",
    "Jio.print_df(partitioned_genetic_regions_df, df_name=\"E coli Genetic Regions partitioned by PSWM Candidates\")"
   ],
   "id": "b2b53d1f99630635",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "E coli Genetic Regions partitioned by PSWM Candidates: 1870127 Row x 4 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "     GeneID  Sequence_Partition  Start  End\n",
      "0  16127995  aacggcagaccaacatca      0   17\n",
      "1  16127995  acggcagaccaacatcaa      1   18\n",
      "2  16127995  cggcagaccaacatcaac      2   19\n",
      "3  16127995  ggcagaccaacatcaact      3   20\n",
      "4  16127995  gcagaccaacatcaactg      4   21\n",
      "5  16127995  cagaccaacatcaactgc      5   22\n",
      "6  16127995  agaccaacatcaactgca      6   23\n",
      "7  16127995  gaccaacatcaactgcaa      7   24\n",
      "8  16127995  accaacatcaactgcaag      8   25\n",
      "9  16127995  ccaacatcaactgcaagc      9   26\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 199
  },
  {
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Evaluate Partitioned Genetic Regions DF with PSWM\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "# Computes the Score of a candidate sequence with the PSWM\n",
    "def evaluate_sequence_by_pswm_df(sequence,pswm_df,vectorized_approach=True):\n",
    "    # Check if the sequence length matches the PSWM's number of columns\n",
    "    if len(sequence) != len(pswm_df.columns):\n",
    "        # Otherwise raise an error\n",
    "        raise ValueError(f\"Sequence length {len(sequence)} does not match PSWM length {len(pswm_df.columns)}.\")\n",
    "\n",
    "    # Check if all unique characters in the sequence are represented in the PSWM index\n",
    "    if not set(sequence).issubset(set(pswm_df.index)):\n",
    "        # Otherwise raise an error\n",
    "        raise ValueError(f\"Sequence contains characters not represented in the PSWM: {set(sequence) - set(pswm_df.index)}\")\n",
    "\n",
    "    # SLOW NESTED LOOP METHOD\n",
    "    if not vectorized_approach:\n",
    "        # Evaluate the sequence with the PSWM\n",
    "        score = 0 # Initalize a score\n",
    "        for i, char in enumerate(sequence):\n",
    "            score += pswm_df.loc[char].iloc[i]\n",
    "\n",
    "    # FAST VECTORIZED METHOD \n",
    "    else:\n",
    "        # Suggested by ChatGPT\n",
    "        # Convert sequence characters to indices using the PSWM index\n",
    "        # This presumes that PSWM's index is ordered as needed by the sequence characters\n",
    "        indices = [pswm_df.index.get_loc(char) for char in sequence]\n",
    "    \n",
    "        # Select the appropriate scores using advanced indexing\n",
    "        scores = pswm_df.values[indices, np.arange(len(sequence))]\n",
    "    \n",
    "        # Sum up the scores\n",
    "        score = np.sum(scores)\n",
    "\n",
    "    # Return the composite score of the sequence\n",
    "    return score\n"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "e71ef016c0d55f46",
   "outputs": [],
   "execution_count": 200
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Original Approach by Jay took 3 seconds and 700 ms\n",
    "\n",
    "# Calculate the Scores for 1% of the DataFrame to Test Validity of the Logic\n",
    "subset_size = int(len(partitioned_genetic_regions_df) * 0.01)\n",
    "subset_df = partitioned_genetic_regions_df.iloc[:subset_size].copy()  \n",
    "\n",
    "\n",
    "# Apply the function to each row of the subset DataFrame\n",
    "subset_df['Score'] = subset_df.apply(\n",
    "    lambda row: evaluate_sequence_by_pswm_df(row['Sequence_Partition'], pswm_df,vectorized_approach=False),\n",
    "    axis='columns'\n",
    ")\n"
   ],
   "id": "88531a57e9f4f75",
   "outputs": [],
   "execution_count": 201
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Vectorized Approach with Numpy as suggested by ChatGPT takes 273 ms\n",
    "\n",
    "# Calculate the Scores for 1% of the DataFrame to Test Validity of the Logic\n",
    "subset_size = int(len(partitioned_genetic_regions_df) * 0.01)\n",
    "subset_df = partitioned_genetic_regions_df.iloc[:subset_size].copy()\n",
    "\n",
    "# Apply the function to each row of the subset DataFrame\n",
    "subset_df['Score'] = subset_df.apply(\n",
    "    lambda row: evaluate_sequence_by_pswm_df(row['Sequence_Partition'], pswm_df,vectorized_approach=True),\n",
    "    axis='columns'\n",
    ")"
   ],
   "id": "f21de6037096d044",
   "outputs": [],
   "execution_count": 202
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Vectorized Approach with Numpy as suggested by ChatGPT takes 273 ms\n",
    "# Original Approach by Jay took 3 seconds and 809 ms\n",
    "\n",
    "# Scored Genetic Regions DF\n",
    "scored_genetic_regions_df = partitioned_genetic_regions_df.copy()\n",
    "\n",
    "# Apply the function to each row of the DataFrame\n",
    "scored_genetic_regions_df['Score'] = scored_genetic_regions_df.apply(\n",
    "    lambda row: evaluate_sequence_by_pswm_df(row['Sequence_Partition'], pswm_df),\n",
    "    axis='columns'\n",
    ")"
   ],
   "id": "ded1c4c87a9d4037",
   "outputs": [],
   "execution_count": 203
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Display the DataFrame with the new 'Score' column\n",
    "print(partitioned_genetic_regions_df)"
   ],
   "id": "d006d31f88be15e8",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "           GeneID  Sequence_Partition  Start  End\n",
      "0        16127995  aacggcagaccaacatca      0   17\n",
      "1        16127995  acggcagaccaacatcaa      1   18\n",
      "2        16127995  cggcagaccaacatcaac      2   19\n",
      "3        16127995  ggcagaccaacatcaact      3   20\n",
      "4        16127995  gcagaccaacatcaactg      4   21\n",
      "...           ...                 ...    ...  ...\n",
      "1870122  16132220  ccgccagagcagaaaata    428  445\n",
      "1870123  16132220  cgccagagcagaaaatat    429  446\n",
      "1870124  16132220  gccagagcagaaaatatt    430  447\n",
      "1870125  16132220  ccagagcagaaaatattg    431  448\n",
      "1870126  16132220  cagagcagaaaatattgg    432  449\n",
      "\n",
      "[1870127 rows x 4 columns]\n"
     ]
    }
   ],
   "execution_count": 204
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Modify the Start and End Positions to Correspond to Upstream and Downstream\n",
    "# Upstream = Negative | Downstream = Positive\n",
    "\n",
    "scored_genetic_regions_df['Start'] -= 400\n",
    "scored_genetic_regions_df['End'] -= 400"
   ],
   "id": "c6dc10164b80b891",
   "outputs": [],
   "execution_count": 205
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "Jio.print_df(scored_genetic_regions_df,'E coli candidate motif regions as evaluated by PSWM')",
   "id": "b53b25aabfd248f4",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "E coli candidate motif regions as evaluated by PSWM: 1870127 Row x 5 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "     GeneID  Sequence_Partition  Start  End      Score\n",
      "0  16127995  aacggcagaccaacatca   -400 -383  -6.009405\n",
      "1  16127995  acggcagaccaacatcaa   -399 -382  -4.967363\n",
      "2  16127995  cggcagaccaacatcaac   -398 -381 -10.219168\n",
      "3  16127995  ggcagaccaacatcaact   -397 -380 -11.167474\n",
      "4  16127995  gcagaccaacatcaactg   -396 -379  -4.301812\n",
      "5  16127995  cagaccaacatcaactgc   -395 -378  -5.766668\n",
      "6  16127995  agaccaacatcaactgca   -394 -377  -0.367716\n",
      "7  16127995  gaccaacatcaactgcaa   -393 -376  -7.400153\n",
      "8  16127995  accaacatcaactgcaag   -392 -375  -9.973552\n",
      "9  16127995  ccaacatcaactgcaagc   -391 -374 -12.817403\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 206
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "#  Gather Top Genes in terms of Motif Consensus\n",
    "# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
    "\n",
    "top_genes_threshold = 30\n",
    "\n",
    "# Group by 'GeneID' and find the row index of the max 'Score' per group\n",
    "idx_max_scores = scored_genetic_regions_df.groupby('GeneID')['Score'].idxmax()\n",
    "\n",
    "# Use the indices to subset the rows with the max 'Score' from the original DataFrame\n",
    "top_genes_df = scored_genetic_regions_df.loc[idx_max_scores]\n",
    "\n",
    "# Sort subsetted rows by 'Score' in descending order and get the top 30\n",
    "top_genes_df = top_genes_df.sort_values(by='Score', ascending=False).head(top_genes_threshold)"
   ],
   "id": "a0690d9ee564f05a",
   "outputs": [],
   "execution_count": 207
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# View the Top 30 Genes\n",
    "Jio.print_df(top_genes_df ,'E coli genes with highest motif candidacy as evaluated by PSWM')"
   ],
   "id": "8907b63696371922",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "E coli genes with highest motif candidacy as evaluated by PSWM: 30 Row x 5 Col\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "           GeneID  Sequence_Partition  Start  End      Score\n",
      "1341103     b3171  tcactgaatttttatgca   -298 -281  14.543283\n",
      "113799   16128258  aaagtgaattttaattca    -47  -30  14.391985\n",
      "1808996  16132076  aaattgaattttaattca    -45  -28  13.804199\n",
      "1369154  16131126  cagatgaatttttatgca   -392 -375  13.234430\n",
      "1421868  16131238  ttagtgattttttatgca    -71  -54  13.089849\n",
      "567057   16129301  tgaatgaataatcatgca   -140 -123  12.701051\n",
      "292111   16128684  ctaatgtatatttatgca   -131 -114  12.536748\n",
      "1129829  16130583  atattgcatttttattca   -268 -251  12.396702\n",
      "1679074  16131795  ttagtgtatttttattca    -67  -50  12.396702\n",
      "13339    16128026  attgtgaattaatatgca    -51  -34  11.908349\n",
      "~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~\n",
      "\n",
      "\n"
     ]
    }
   ],
   "execution_count": 208
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Save the Results of the PSWM Evaluation\n",
    "if save_file:\n",
    "\n",
    "    # Establish the Result DFs to Save\n",
    "    result_dfs = {\n",
    "        'Top30': top_genes_df,\n",
    "        'All': scored_genetic_regions_df,\n",
    "    }\n",
    "    \n",
    "    # Save the Result DFs\n",
    "    for subset, result_df in result_dfs.items():\n",
    "        # Save the DF in the specified file format\n",
    "        Jio.df_to_file(\n",
    "            df=result_df,\n",
    "            file_name=f\"Gene_Binding_Site_Evaluations_{subset}\",\n",
    "            file_format=output_format\n",
    "        )\n"
   ],
   "id": "b23aedc3c3017c6d",
   "outputs": [],
   "execution_count": 209
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ~~~~~~~~~~~~~~~\n",
    "#  End of Script\n",
    "# ~~~~~~~~~~~~~~~\n"
   ],
   "id": "b62b2d3ac212a603",
   "outputs": [],
   "execution_count": 210
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
